### PR TITLE
JP-3437: Address alternating column noise for NIRSpec IRS2

### LINKED ISSUE #7995: Alternating column noise not corrected in refpix step
_Issue [JP-3437](https://jira.stsci.edu/browse/JP-3437) was created on JIRA by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke):_

We have had a couple of user reports of some low-level detector noise appearing in NIRSpec rate files that looks like odd-even row striping across part of the array (rows in the science orientation; these are columns in the detector orientation).

We think this is alternating column noise (ACN), due to the two amplifiers (one for odd and one for even columns) in an output having slightly different offsets. Usually the amplifiers should be close, but they may drift over the course of an exposure or be impacted by a cosmic ray hit.

We have tried turning on and off the odd_even_columns parameter in the refpix step, but it makes very little difference to the data when it is set to True or False.  Is this parameter working as intended? If so, but it does not apply to this use case, is any other mitigation possible?

### PR DESCRIPTION
**Pull request by melanieclarke**
Resolves [JP-3437](https://jira.stsci.edu/browse/JP-3437)

Closes #7995

This PR adapts traditional readout handling for NIRSpec IRS2 mode, to subtract mean reference pixel offsets by amplifier and column parity.  The offset subtraction is done before IRS2 handling for interleaved reference pixels.

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [x] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by melanieclarke** [comment]
I think this is ready for review now, if someone can run regression tests for me.  I expect changes only for NIRSpec IRS2 data, downstream of the refpix step.  In local runs of the regression tests, I see improvements to alternating column noise for IRS2 data, compared to the truth files.

**Comment by tapastro** [comment]
Started a regression set: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1116/

**Comment by melanieclarke** [comment]
It looks like there are some unrelated failures in the regression tests, for MIRI header keywords and a couple other things.  The NIRSpec changes are as expected.

**Comment by hbushouse** [comment]
Started another regtest run here: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1127  Truth files have been updated for recent ref file updates, so unrelated failures should go away now.

**Review by hbushouse — COMMENTED** [review]
Code updates look good to me. Do we need to make updates in the refpix step docs (RTD) for this change to the way IRS2 data are handled?

**Comment by melanieclarke** [comment]
Thanks @hbushouse.  I'll take a look at the docs.

I also need to take another look at how even and odd interleaved pixels are identified, to address a comment in the ticket.  I'll put this PR back to draft status while I work on it.

**Comment by hbushouse** [comment]
All regtest failures in the latest run are now confined to NRS tests and appear to be small differences in array values, as expected.

**Comment by melanieclarke** [comment]
Apologies, I'm setting this back to draft again, because testing shows that this change with current reference files sometimes introduces alternating column noise where it didn't exist before.  This needs more study before merging this change.

**Comment by melanieclarke** [comment]
I think this is ready for review again.  

The current state is that the code changes are complete, but they are off by default.  With the current reference files, these changes helps reduce ACN in some cases, but sometimes introduce ACN when odd_even_columns=True.  

I have a secondary need for the mean value subtraction with odd_even_columns=False, for generating new superbias reference files, so it would be helpful to the NIRSpec team to have the option available, even if it is not used by default.

File: jwst/refpix/reference_pixels.py [hunk 1]
```diff
@@ -481,18 +506,128 @@ def __init__(self, input_model,
...
+        """
+        Make an odd pixel mask for IRS2 mode.
+
+        The even pixel mask can be generated by inverting the odd mask.
+
+        Parameters
+        ----------
+        input_model : DataModel
+            Input model containing data to mask.
+        scipix_n_default: int, optional
```
Comment (hbushouse) [inline jwst/refpix/reference_pixels.py:529]:
```suggestion
        scipix_n_default : int, optional
```

File: jwst/refpix/reference_pixels.py [hunk 2]
```diff
@@ -481,18 +506,128 @@ def __init__(self, input_model,
...
+            Number of regular samples before stepping out to collect
+            reference samples.
+        refpix_r_default: int, optional
```
Comment (hbushouse) [inline jwst/refpix/reference_pixels.py:532]:
```suggestion
        refpix_r_default : int, optional
```

File: jwst/refpix/reference_pixels.py [hunk 3]
```diff
@@ -481,18 +506,128 @@ def __init__(self, input_model,
...
+        For traditional readouts, odd pixels correspond to odd-numbered
+        rows (first, third, fifth, etc.), which are even array indices.
 
+        For IRS2 mode, science and traditional reference pixels have
+        the same parity, but interleaved pixels come in two pairs. The
+        first two are odd and the second two are even.
+
+        Parameters
+        ----------
         group: NDArray
```
Comment (hbushouse) [inline jwst/refpix/reference_pixels.py:631]:
```suggestion
        group : NDArray
```

File: jwst/refpix/reference_pixels.py [hunk 4]
```diff
@@ -503,8 +638,8 @@ def collect_odd_refpixels(self, group, amplifier, top_or_bottom):
             String corresponding to whether top or bottom reference pixels
             are bing processed
 
-        Returns:
-
+        Returns
+        -------
         oddref: NDArray
```
Comment (hbushouse) [inline jwst/refpix/reference_pixels.py:643]:
```suggestion
        oddref : NDArray
```

File: jwst/refpix/reference_pixels.py [hunk 5]
```diff
@@ -513,18 +648,31 @@ def collect_odd_refpixels(self, group, amplifier, top_or_bottom):
...
+        rows (second, fourth, sixth, etc.), which are odd array indices.
+
+        For IRS2 mode, science and traditional reference pixels have
+        the same parity, but interleaved pixels come in two pairs. The
+        first two are odd and the second two are even.
+
+        Parameters
+        ----------
 
         group: NDArray
```
Comment (hbushouse) [inline jwst/refpix/reference_pixels.py:677]:
```suggestion
        group : NDArray
```

File: jwst/refpix/reference_pixels.py [hunk 6]
```diff
@@ -536,23 +684,28 @@ def collect_even_refpixels(self, group, amplifier, top_or_bottom):
...
-        Returns:
-
-        oddref: NDArray
-            Array containing all the odd reference pixels
-
-        odddq: NDArray
-            Array containing all the odd dq values for those reference pixels
+        Returns
+        -------
+        evenref: NDArray
```
Comment (hbushouse) [inline jwst/refpix/reference_pixels.py:689]:
```suggestion
        evenref : NDArray
```

File: jwst/refpix/reference_pixels.py [hunk 7]
```diff
@@ -536,23 +684,28 @@ def collect_even_refpixels(self, group, amplifier, top_or_bottom):
...
+            Array containing all the even reference pixels
 
+        evendq: NDArray
```
Comment (hbushouse) [inline jwst/refpix/reference_pixels.py:692]:
```suggestion
        evendq : NDArray
```

**Review by hbushouse — COMMENTED** [review]
Just a few super minor formatting suggestions.

**Comment by hbushouse** [comment]
Another regtest run started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1243

**Comment by melanieclarke** [comment]
> Just a few super minor formatting suggestions.

Thanks!  I committed your suggestions.

**Comment by hbushouse** [comment]
@melanieclarke All the differences in the latest regtest run are unrelated and due to the recent addition of a new keyword in stdatamodels. Is it expected that we won't see any changes to NIRSpec images, because the changes included here are off by default?

**Comment by melanieclarke** [comment]
Yes, that's right. There should be no changes to default reduction.  We'll see if we can/should turn it on by default when current IRS2 reference file investigations are done.

**Comment by hbushouse** [comment]
OK, so in that case I think this is ready to merge.

**Comment by melanieclarke** [comment]
Thanks very much for reviewing, @hbushouse!

**Comment by melanieclarke** [comment]
@hbushouse - Just checking -- it looks like this was scheduled to auto-merge but did not actually go through.  Is it waiting for something else?

**Review by hbushouse — APPROVED** [review]

**Comment by hbushouse** [comment]
> @hbushouse - Just checking -- it looks like this was scheduled to auto-merge but did not actually go through. Is it waiting for something else?

When I try to merge I get an error message saying "Merge attempt failed. Head branch is out of date. Review and try the merge again."  Never seen that before!  Not even sure what "head branch is out of date" might mean. Perhaps do a rebase of your local PR branch and trying pushing one last time?

**Comment by melanieclarke** [comment]
That's odd.  I rebased from main and pushed again, but I still see a "Head branch was modified" message above.  Maybe it needs you to approve again?

**Comment by hbushouse** [comment]
Yay, successfully merged this time!

**Pull request merged**